libminifi/src/utils/RegexUtils.cpp

```diff
@@ -0,0 +1,168 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "utils/RegexUtils.h"
+#include "Exception.h"
+#include <iostream>
+#include <vector>
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace utils {
+
+Regex::Regex() : Regex::Regex("") {}
+
+Regex::Regex(const std::string &value) : Regex::Regex(value, {}) {}
+
+Regex::Regex(const std::string &value,
+                           const std::vector<Regex::Mode> &mode)
+    : regexStr(value),
+      valid(false) {
+  if (regexStr.empty())
+    return;
+
+  // Create regex mode
+#ifdef NO_MORE_REGFREEE
+  regex_mode = std::regex_constants::ECMAScript;
+#else
+  regex_mode = REG_EXTENDED;
+#endif
+  for (const auto m : mode) {
+    switch (m) {
+      case Mode::ICASE:
+#ifdef NO_MORE_REGFREEE
+        regex_mode |= std::regex_constants::icase;
+#else
+        regex_mode |= REG_ICASE;
+#endif
+        break;
+    }
+  }
+  // Compile
+#ifdef NO_MORE_REGFREEE
+  try {
+    compiledRegex = std::regex(regexStr, regex_mode);
+    valid = true;
+  } catch (const std::regex_error &e) {
+    throw Exception(REGEX_EXCEPTION, e.what());
+  }
+#else
+  int err_code = regcomp(&compiledRegex, regexStr.c_str(), regex_mode);
+  if (err_code) {
+    const size_t sz = regerror(err_code, &compiledRegex, nullptr, 0);
+    std::vector<char> msg(sz);
+    regerror(err_code, &compiledRegex, msg.data(), msg.size());
+    throw Exception(REGEX_EXCEPTION, std::string(msg.begin(), msg.end()));
+  }
+  valid = true;
+  int maxGroups = std::count(regexStr.begin(), regexStr.end(), '(') + 1;
+  matches.resize(maxGroups);
+#endif
+}
+
+Regex::Regex(Regex&& other)
+#ifndef NO_MORE_REGFREEE
+  : valid(false),
+    regex_mode(REG_EXTENDED)
+#endif
+{
+  *this = std::move(other);
+}
+
+Regex& Regex::operator=(Regex&& other) {
+  if (this == &other) {
+    return *this;
+  }
+
+  pat = std::move(other.pat);
+  suffix = std::move(other.suffix);
+  regexStr = std::move(other.regexStr);
+  results = std::move(other.results);
+  valid = other.valid;
+  other.valid = false;
+#ifdef NO_MORE_REGFREEE
+  compiledRegex = std::move(other.compiledRegex);
+  regex_mode = other.regex_mode;
+  matches = std::move(other.matches);
+#else
+  if (valid)
```

Comment:
This should depend on the original valid value, but it depends on other's valid value.